### PR TITLE
feat(linter): convert-tslint-to-eslint schematic

### LINKED ISSUE #4320: Migration to ESlint
I have a nx workspace already configured with TSlint and I've migrated to nx 11 recently.
How do I migrate to ESLint? [angular-eslint](https://github.com/angular-eslint/angular-eslint#usage-with-nx-monorepos) instruct us to follow nx path in order to migrate properly. Which path would it be? I couldn't find it anywhere.

### Environment
``` Node : 12.14.0
  OS   : linux x64
  npm  : 6.13.4
  
  nx : Not Found
  @nrwl/angular : 11.0.4
  @nrwl/cli : 11.0.4
  @nrwl/cypress : 11.0.4
  @nrwl/devkit : 11.0.4
  @nrwl/eslint-plugin-nx : Not Found
  @nrwl/express : Not Found
  @nrwl/jest : 11.0.4
  @nrwl/linter : 11.0.4
  @nrwl/nest : Not Found
  @nrwl/next : Not Found
  @nrwl/node : Not Found
  @nrwl/react : Not Found
  @nrwl/schematics : Not Found
  @nrwl/tao : 11.0.4
  @nrwl/web : Not Found
  @nrwl/workspace : 11.0.4
  typescript : 4.0.5```

### PR DESCRIPTION
**Pull request by JamesHenry**
## Current Behavior

Nx supports creating Angular (and some others like NestJS) projects with both TSLint and ESLint as the linter, but with TSLint being EOL at the end of 2020, there is a need to be able to migrate existing TSLint projects over to ESLint.

This is not currently available in Nx.

## Expected Behavior

Nx provides a schematic/generator to assist with the conversion from TSLint to ESLint.

## Related Issue(s)

Fixes #4320

### CODE REVIEW

**Comment by JamesHenry** [comment]
I got a dep-check failure, and have addressed it by adding the `ignore` mapping in `missing.ts`, but not 100% sure if that was the right way to deal with it. All the packages in question are automatically installed by schematics/generators

File: packages/angular/src/schematics/convert-tslint-to-eslint/convert.ts [hunk 1]
```diff
@@ -0,0 +1,840 @@
...
+      '\nYou will need to decide on how to handle the above manually, but everything else has been handled for you automatically.\n'
+    );
+  }
+}
+
+function likelyContainsTSLintComment(fileContent: string): boolean {
+  return fileContent.includes('tslint:');
+}
+
+function convertTSLintDisableCommentsForProject(projectName: string): Rule {
```
Comment (FrozenPandaz) [inline packages/angular/src/schematics/convert-tslint-to-eslint/convert.ts:800]: You could use this method which will also not visit ignored directories and files:

https://github.com/nrwl/nx/blob/master/packages/workspace/src/utils/rules/visit-not-ignored-files.ts#L10

File: packages/angular/src/schematics/convert-tslint-to-eslint/convert.ts [hunk 2]
```diff
@@ -0,0 +1,840 @@
...
+       */
+      removeExtendsFromProjectTSLintConfigBeforeConverting(
+        projectRoot,
+        projectTSLintJsonPath
+      ),
+      /**
+       * The conversion logic needs to be different for Angular projects vs e2e projects
+       * vs other project types which could use TSLint (such as NestJS).
+       */
+      isE2EProject(schema.project, projectConfig)
```
Comment (FrozenPandaz) [inline packages/angular/src/schematics/convert-tslint-to-eslint/convert.ts:129]: It seems weird to me that we have logic within `@nrwl/angular` to handle other `@nrwl/*` things.

I think this generator belongs more in the `@nrwl/linter` package. I also think we may have a lot of utils in that package which will make it easier to work with lint configurations.

Yes, it will mean the user will have to install `@nrwl/linter` but maybe we can make that work to our advantage? :thinking:

I'm thinking we could make `convert-tslint-to-eslint` a dependency of `@nrwl/linter` which would mean we don't need an installation step in this generator.

What do you think?

Comment (FrozenPandaz) [inline packages/angular/src/schematics/convert-tslint-to-eslint/convert.ts:129]: I see that in this case, we are adding the `cypress` eslint rules to the config. I think the behavior should be something more along the lines of `isCypressProject`.

File: packages/angular/src/schematics/convert-tslint-to-eslint/convert.ts [hunk 3]
```diff
@@ -0,0 +1,840 @@
...
+    const rootESLintrcJsonPath = join(
+      normalize(tree.root.path),
+      '.eslintrc.json'
+    );
+
+    return chain([
+      /**
+       * Handle adding ESLint and related dependencies and configuration
+       */
+      addAllConfigFilesAndCoreDependencies(projectRoot),
```
Comment (FrozenPandaz) [inline packages/angular/src/schematics/convert-tslint-to-eslint/convert.ts:117]: I don't know where the issue is but what I'm seeing on `nx-examples` is that the depConstraints didn't get migrated correctly.

I have `depConstraints` defined in `/tslint.json` but the `/.eslintrc.json` has the `depConstraints`.

File: packages/angular/src/schematics/convert-tslint-to-eslint/convert.ts [hunk 4]
```diff
@@ -0,0 +1,840 @@
...
+    delete codeRules[ruleName];
+  });
+
+  return {
+    codeRules,
+    templateRules,
+  };
+}
+
+function dedupeEnvAgainstConfigs(
```
Comment (FrozenPandaz) [inline packages/angular/src/schematics/convert-tslint-to-eslint/convert.ts:700]: These `dedupe` methods seem like duplicates of what we did during the `presets` migration. I think moving this generator to `@nrwl/linter` should enable us to share that code.

**Review by FrozenPandaz — DISMISSED** [review]
I didn't go through every little detail but I think maybe moving it to `@nrwl/linter` would be helpful?

File: packages/angular/src/schematics/convert-tslint-to-eslint/convert.ts (hunk 2 above)
Comment (JamesHenry) [reply packages/angular/src/schematics/convert-tslint-to-eslint/convert.ts:129]: Why would we not need an installation step in the generator, sorry? It relies on eslint-plugin-nx, not linter, so I don't think anything about it works will change based on what you are suggesting?

Comment (JamesHenry) [reply packages/angular/src/schematics/convert-tslint-to-eslint/convert.ts:129]: Good catch - I think my brain sometimes like to pretend it's the only option 😅

File: packages/angular/src/schematics/convert-tslint-to-eslint/convert.ts (hunk 3 above)
Comment (JamesHenry) [reply packages/angular/src/schematics/convert-tslint-to-eslint/convert.ts:117]: Please could you try and add a failing unit test to reproduce? The conversion logic for the nx-module-boundaries rule has its own tests in this PR

File: packages/angular/src/schematics/convert-tslint-to-eslint/convert.ts (hunk 4 above)
Comment (JamesHenry) [reply packages/angular/src/schematics/convert-tslint-to-eslint/convert.ts:700]: These are actually an improved version I think, but I'll need to go back and double check

File: packages/angular/src/schematics/convert-tslint-to-eslint/convert.ts [hunk 5]
```diff
@@ -0,0 +1,840 @@
...
+    /**
+     * We are not able to support --dry-run in this schematic, because we need to (potentially) both install
+     * and use the @nrwl/eslint-plugin-nx package within the same execution.
+     *
+     * This is a worthwhile trade-off to avoid the need to force the user to manually run an install before
+     * invoking the schematic, and the dry-run output doesn't offer a ton of value for this use-case anyway.
+     */
+    if (
+      process.argv.includes('--dry-run') ||
+      process.argv.includes('--dryRun')
```
[the remark this answers is not in the document]
Comment (JamesHenry) [reply packages/angular/src/schematics/convert-tslint-to-eslint/convert.ts:57]: Ah thanks

File: packages/angular/src/schematics/convert-tslint-to-eslint/convert.ts (hunk 1 above)
Comment (JamesHenry) [reply packages/angular/src/schematics/convert-tslint-to-eslint/convert.ts:800]: Cool thanks

Comment (JamesHenry) [reply packages/angular/src/schematics/convert-tslint-to-eslint/convert.ts:800]: I've added some previously missing test coverage for this logic too

**Comment by JamesHenry** [comment]
Migrated the schematic from the angular package to the linter package, as per discussion with Jason

File: packages/angular/src/schematics/convert-tslint-to-eslint/convert.ts (hunk 3 above)
Comment (JamesHenry) [reply packages/angular/src/schematics/convert-tslint-to-eslint/convert.ts]: Ok actually I had forgotten my unit tests were already baed on specifically that repo so looks like unit testing won't reproduce it. I'll try and reproduce with a local build on the repo itself

> I have depConstraints defined in /tslint.json but the /.eslintrc.json has the depConstraints.

I'm guessing this has a typo in it? And that the eslintrc does _not_ have them?

Comment (JamesHenry) [reply packages/angular/src/schematics/convert-tslint-to-eslint/convert.ts]: Have reproduced locally on the nx-examples repo 👍 will dig into that issue

Comment (JamesHenry) [reply packages/angular/src/schematics/convert-tslint-to-eslint/convert.ts]: Resolved

**Comment by JamesHenry** [comment]
Reconverted to DRAFT for now - want to merge in angular-eslint v1 update so that everything is nice a clean, and need to do a bit more work on extracting custom rules in user's root tslint.json and converting them to the right location in project configs

**Comment by vsavkin** [comment]
@JamesHenry could you rebase this PR on master as soon as you can? Thank you!

**Comment by s4m0r4m4** [comment]
Hello I'm just checking on this one. Looking to do a transition to ESLint in the next week or two, and it'd be great to utilize this schematic even if it just gets me 75% of the way there.

**Comment by JamesHenry** [comment]
@s4m0r4m4 I'm working on it more this week

**Comment by JamesHenry** [comment]
This is superseded by the new PR here which is feature complete: https://github.com/nrwl/nx/pull/4943

Thanks all! 🙏
